core/gallery/cache.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		log.Warn("failed to read local index. Need to refetch index from remote", zap.Error(err))
+	}
+
+	version := ""
+	if localIndex != nil {
+		version, err = c.storage.getVersion()
+		if err != nil {
+			log.Warn("failed to read local version. Need to refetch version from remote", zap.Error(err))
+		}
+	}
+
+	index, newVersion, err := c.downloadGalleryIndex(timeoutInSeconds, version, true)
+	if err == nil {
+		c.storage.save(index, newVersion)
+
+		return getManifestByDownloadLink(index, downloadLink)
+	}
+
+	if errors.Is(err, ErrNotModified) {
+		manifest, err := getManifestByDownloadLink(localIndex, downloadLink)
+		if err != nil {
+			return nil, err
+		}
+		return manifest, nil
+	}
+
+	if localIndex != nil {
+		log.Warn("failed to download index from remote. Returning local index", zap.Error(err))
+		manifest, err := getManifestByDownloadLink(localIndex, downloadLink)
+		if err != nil {
+			return nil, err
+		}
+		return manifest, nil
+	}
+
+	return nil, err
+}
+
+type cacheStorage interface {
+	getIndex() (*pb.RpcGalleryDownloadIndexResponse, error)
+	getVersion() (string, error)
+	save(index *pb.RpcGalleryDownloadIndexResponse, version string)
+}
+
+type storage struct {
+	versionPath, indexPath string
+}
+
+func (s *storage) getIndex() (*pb.RpcGalleryDownloadIndexResponse, error) {
+	rawData, err := os.ReadFile(s.indexPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read local gallery index: %w", err)
+	}
+
+	index := &pb.RpcGalleryDownloadIndexResponse{}
+	if err = proto.Unmarshal(rawData, index); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal local gallery index: %w", err)
+	}
+	return index, nil
+}
+
+func (s *storage) getVersion() (string, error) {
+	rawData, err := os.ReadFile(s.versionPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read local gallery index version: %w", err)
+	}
+	return string(rawData), nil
+}
+
+// TODO: Maybe we should save info to files async?
+func (s *storage) save(index *pb.RpcGalleryDownloadIndexResponse, version string) {
+	data, err := proto.Marshal(index)
+	if err != nil {
+		log.Error("failed to marshal local gallery index", zap.Error(err))
+		return
+	}
+
+	if err = os.WriteFile(s.indexPath, data, 0600); err != nil {
+		log.Error("failed to save local gallery index", zap.Error(err))
+		return
+	}
+
+	if err = os.WriteFile(s.versionPath, []byte(version), 0600); err != nil {
+		log.Error("failed to save local gallery version", zap.Error(err))
+	}
+}
+
+func (c *cache) downloadGalleryIndex(
+	timeoutInSeconds int, // timeout to wait for HTTP response
+	version string, // Etag of gallery index, that allows to fetch index faster
+	withManifestValidation bool, // a flag that indicates that every manifest should be validated
+) (response *pb.RpcGalleryDownloadIndexResponse, newVersion string, err error) {
+	raw, newVersion, err := getRawJson(c.indexURL, timeoutInSeconds, version)
+	if err != nil {
+		if errors.Is(err, ErrNotModified) {
```

Comment:
This check is not needed, because it's implemented in upper function afaiu? 🧐

Reply:
It is actually needed, as in case requested resource is not actually modified we do not wrap it into ErrDownloadIndex, that is translated into RPC return code